src/script/common/helper.h

```diff
@@ -24,34 +24,231 @@ extern "C" {
 #include <lauxlib.h>
 }
 
+#include <vector>
+#include <string>
+#include "irrlichttypes_extrabloated.h"
+
 class LuaHelper
 {
+private:
+#define MAKE_SINGLE_ARG(name, type)									\
+	template <typename T> struct name : public std::false_type {	\
+		typedef T value_type;										\
+	};																\
+	template <typename T>											\
+	struct name<type<T>> : public std::true_type {					\
+		typedef T value_type;										\
+	};
+
+#define MAKE_MULTI_ARG(name, type)									\
+	template <typename T> struct name : public std::false_type {	\
+		typedef T value_type;										\
+	};																\
+	template <typename T, typename... Args>							\
+	struct name<type<T, Args...>> : public std::true_type {			\
+		typedef T value_type;										\
+	};
+
+	MAKE_SINGLE_ARG(is_vector2d, core::vector2d)
+	MAKE_SINGLE_ARG(is_vector3d, core::vector3d)
+
+	MAKE_MULTI_ARG(is_vector, std::vector)
+
+#undef MAKE_SINGLE_ARG
+#undef MAKE_MULTI_ARG
+
+	// Specializations are outside of the class
+	template <typename T> struct is_integer : public std::false_type {
+		typedef T value_type;
+	};
+
+#define SPECIAL_TEMPLATE(name) template <typename T, \
+	typename std::enable_if<name<T>::value>::type* = nullptr>
+#define DISABLE_TEMPLATE template <typename T, typename std::enable_if< \
+		!is_integer<T>::value &&	\
+		!is_vector<T>::value &&		\
+		!is_vector2d<T>::value &&	\
+		!is_vector3d<T>::value		\
+	>::type* = nullptr>
+
+#define TEMPLATES(function)					\
+	DISABLE_TEMPLATE function;				\
+	SPECIAL_TEMPLATE(is_integer) function;	\
+	SPECIAL_TEMPLATE(is_vector2d) function;	\
+	SPECIAL_TEMPLATE(is_vector3d) function;	\
+	SPECIAL_TEMPLATE(is_vector) function
+
 protected:
-	static bool isNaN(lua_State *L, int idx);
+	/**
+	 * Displays an error message appropriate for invalid or unexpected types
+	 *
+	 * @param L Lua state
+	 * @param index Index on the Lua stack that was erroneous
+	 * @param name Name of the expected type
+	 * @param type Expected Lua type
+	 */
+	static void invalidType(lua_State *L, int index, const std::string &name, int type);
+
+	/**
+	 * Returns whether the given value on the Lua stack is of the specified type.
+	 */
+	TEMPLATES(static std::string getTypeName(bool plural = false));
```

Comment:
should return `const char*`
(also the description above is wrong)

Reply:
`std::string` is intentional; for instance:

```
SPECIAL_TEMPLATE(LuaHelper::is_vector) std::string LuaHelper::getTypeName(
	bool plural)
{
	std::string start;
	if (plural)
		start = "arrays of ";
	else
		start = "array of ";
	return start + getTypeName<typename T::value_type>(true);
}
```
This allows `getTypeName<std::vector<core::vector2di<u32>>>()` to return `"array of 2D positions/vectors"`.  Can't be done with `const char *` or else I would have used that.

Description is wrong, fixed.